stateproof/builder.go

```diff
@@ -0,0 +1,470 @@
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := spw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	builderForRound, ok := spw.builders[sfa.Round]
+	if !ok {
+		latest := spw.ledger.Latest()
+		latestHdr, err := spw.ledger.BlockHdr(latest)
+		if err != nil {
+			// The latest block in the ledger should never disappear, so this should never happen
+			return network.Disconnect, err
+		}
+
+		if sfa.Round < latestHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound {
+			// Already have a complete state proof in ledger.
+			// Ignore this sig.
+			return network.Ignore, nil
+		}
+
+		// The sig should be for a round which is a multiple of StateProofInterval
+		// using the latestHdr protocol, since changing StateProofInterval is not supported
+		proto := config.Consensus[latestHdr.CurrentProtocol]
+
+		// proto.StateProofInterval is not expected to be 0 after passing StateProofNextRound
+		// checking anyway, otherwise will panic
+		if proto.StateProofInterval == 0 {
+			return network.Disconnect, fmt.Errorf("handleSig: StateProofInterval is 0 for round %d", latest)
+		}
+
+		if uint64(sfa.Round)%proto.StateProofInterval != 0 {
+			// reject the sig for the round which is not a multiple of the interval
+			// Disconnect: should not be sending a sig for this round
+			return network.Disconnect, fmt.Errorf("handleSig: round %d is not a multiple of SP interval %d",
+				sfa.Round, proto.StateProofInterval)
+		}
+
+		builderForRound, err = spw.makeBuilderForRound(sfa.Round)
+		if err != nil {
+			// Should not disconnect this peer, since this is a fault of the relay
+			// The peer could have other signatures what the relay is interested in
+			return network.Ignore, err
+		}
+		spw.builders[sfa.Round] = builderForRound
+	}
+
+	pos, ok := builderForRound.voters.AddrToPos[sfa.SignerAddress]
+	if !ok {
+		return network.Disconnect, fmt.Errorf("handleSig: %v not in participants for %d", sfa.SignerAddress, sfa.Round)
+	}
+
+	if isPresent, err := builderForRound.Present(pos); err != nil || isPresent {
+		// Signature already part of the builderForRound, ignore.
+		return network.Ignore, nil
+	}
+
+	if err := builderForRound.IsValid(pos, &sfa.Sig, true); err != nil {
+		return network.Disconnect, err
+	}
+
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return addPendingSig(tx, sfa.Round, pendingSig{
+			signer:       sfa.SignerAddress,
+			sig:          sfa.Sig,
+			fromThisNode: sender == nil,
+		})
+	})
+	if err != nil {
+		return network.Ignore, err
```

Comment:
why not network.Disconnect?

Reply:
In this specific case, the relay tried to insert the signature into the table, and it failed. As I see it, a relay wants to disconnect a peer only if a peer is malicious . in this case the peer hasn't done anything wrong.

any thoughts on this strategy?

Reply:
agreed — don't disconnect if there was a local DB error